### PR TITLE
add EXPLAIN ANALYZE statement

### PR DESCRIPTION
**Pull request by chaudum**

### CODE REVIEW

File: blackbox/docs/sql/statements/explain.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@
 ``EXPLAIN``
 ===========
 
-Explain the plan for a given statement.
+Explain and analyze the plan for a given statement.
```
Comment (mxm) [inline blackbox/docs/sql/statements/explain.rst]: explain **or** analyze?

File: blackbox/docs/sql/statements/explain.rst [hunk 2]
```diff
@@ -25,16 +25,22 @@ The ``EXPLAIN`` command displays the execution plan that the planner generates
 for the supplied statement. The plan is returned as a nested object containing
 the plan tree.
 
+When issuing ``EXPLAIN ANALYZE`` the plan of the statement is also executed. In
+this case, timings of the different phases of the plan are returned instead of
+the plan itself.
```
Comment (mxm) [inline blackbox/docs/sql/statements/explain.rst]: Maybe rephrase to: When issuing ``EXPLAIN ANALYZE`` the plan of the statement is executed and timings of the different phases of the plan are returned.

File: core/src/main/java/io/crate/profile/TimerToken.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.profile;
+
+public interface TimerToken {
```
Comment (mxm) [inline core/src/main/java/io/crate/profile/TimerToken.java]: I don't understand the name. Maybe consider renaming to `TimeMeasurable`?

File: sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java [hunk 4]
```diff
@@ -58,25 +87,130 @@ public void execute(DependencyCarrier executor,
...
+                    plan, executor, plannerContext, params, valuesBySubQuery);
+
+                resultReceiver.completionFuture()
+                    .whenComplete(createResultConsumer(executor, consumer, jobId, timerToken, operationTree));
+
+                LogicalPlanner.executeNodeOpTree(executor, jobId, noopRowConsumer, context.enabled(), operationTree);
             } else {
-                consumer.accept(null, new UnsupportedOperationException("EXPLAIN not supported for " + subPlan));
-                return;
+                consumer.accept(null, new UnsupportedOperationException("EXPLAIN ANALYZE not supported for " + plan));
```
Comment (mxm) [inline sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java]: Maybe throw a more descriptive error message for the user?

File: sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java [hunk 5]
```diff
@@ -151,4 +151,23 @@ public void testFailureClosesAllSubContexts() throws Exception {
...
+        builder.addSubContext(ctx1);
+        AbstractExecutionSubContextTest.TestingExecutionSubContext ctx2 = new AbstractExecutionSubContextTest.TestingExecutionSubContext(2);
+        builder.addSubContext(ctx2);
+        JobExecutionContext jobExecutionContext = builder.build();
+
+        jobExecutionContext.start();
+        // kill because the testing subcontexts would run infinitely
+        jobExecutionContext.kill();
+        assertTrue(jobExecutionContext.executionTimes().containsKey("1-TestingExecutionSubContext"));
+        assertTrue(jobExecutionContext.executionTimes().containsKey("2-TestingExecutionSubContext"));
```
Comment (mxm) [inline sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java]: Check for values > 0 here?

File: core/src/main/java/io/crate/profile/ProfilingContext.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    public TimerToken createAndStartTimer(String name) {
+        TimerToken timer = createTimer(name);
+        timer.start();
+        return timer;
+    }
+
+    public void stopAndAddTimer(TimerToken token) {
+        token.stop();
+        if (enabled) {
+            map.put(token.name(), token.durationNanos() / 1_000_000L);
```
Comment (mxm) [inline core/src/main/java/io/crate/profile/ProfilingContext.java]: We use nano time but the results are in milliseconds? Why not use `System.currentTimeMillis()` then?

File: sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+@Singleton
+public class TransportCollectProfileNodeAction implements NodeAction<NodeCollectProfileRequest, NodeCollectProfileResponse> {
```
Comment (mxm) [inline sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java]: Could we add some comment about what this class does?

File: core/src/main/java/io/crate/profile/ProfilingContext.java (hunk 6 above)
Comment (chaudum) [reply core/src/main/java/io/crate/profile/ProfilingContext.java]: Taking the more precise start and stop times still gives more accurate durations even though they are converted to `ms` thereafter, especially for very short durations.

**Comment by chaudum** [comment]
@mxm addressed your comments
@seut since @matriv is not available, would you like to take a look?

File: sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ *   * In the first step the profiled statement is executed and the measurements are written to the JobExecutionContext.
+ *   * In the second step, these measurements are collected from the JobExecutionContexts on each node. This transport
+ *     action is used to perform the collect operation.
+ *
+ */
+@Singleton
+public class TransportCollectProfileNodeAction implements NodeAction<NodeCollectProfileRequest, NodeCollectProfileResponse> {
+
+    private static final String TRANSPORT_ACTION = "crate/sql/node/collectprofile";
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java]: minor: I'd prefer `crate/sql/node/collect/profile` or `crate/sql/node/profile/collect`

File: sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    private static final String EXECUTOR = ThreadPool.Names.SEARCH;
+
+    private final Transports transports;
+    private final JobContextService jobContextService;
+
+    @Inject
+    public TransportCollectProfileNodeAction(TransportService transportService,
+                                    Transports transports,
+                                    JobContextService jobContextService) {
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/profile/TransportCollectProfileNodeAction.java]: pls fix indention

File: sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.transport.TransportResponse;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+public class NodeCollectProfileResponse extends TransportResponse {
+
+    private Map<String, Long> durationByPhase = Collections.emptyMap();
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java]: what is saved here as the String key?  its somehow not obvious to me... I guess it won't  be the phase name as one node can run multiple phases with the same name inside one plan...

Comment (chaudum) [reply sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java]: hm, you're right, the naming of the map is not 100 percent accurate. the keys are `phaseId + phaseName` so they are unique per job execution context. however, you could theoretically use any string as key for the timing.

Comment (chaudum) [reply sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java]: what about `durationByPhaseIdent` as property name?

Comment (chaudum) [reply sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java]: strictly speaking it is even the id + name of the context ;)

Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/profile/NodeCollectProfileResponse.java]: yes that naming would help. if its not documented/obvious (maybe i missed something) it may make sense to add a short comment about that somewhere

**Comment by chaudum** [comment]
@seut addressed your comments

File: sql/src/main/java/io/crate/execution/jobs/JobExecutionContext.java [hunk 11]
```diff
@@ -207,14 +248,42 @@ public long kill() {
...
+    }
+
+    public CompletableFuture<Map<String, Long>> finishProfiling() {
+        if (profiler.enabled()) {
+            assert profilingFuture != null : "profilingFuture must not be null";
+            return profilingFuture.whenComplete((o, t) -> close());
+        }
+        // sanity check
+        IllegalStateException stateException = new IllegalStateException(String.format(Locale.ENGLISH, "Tried to finish profiling job [id=%s], but profiling is not enabled.", jobId));
+        return CompletableFutures.failedFuture(stateException);
```
Comment (mxm) [inline sql/src/main/java/io/crate/execution/jobs/JobExecutionContext.java]: I would write this the other way around, first handling the error case, then the regular case. When I see an if condition at the beginning of a function, I assume it handles a special case rather than the expected case.

File: sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java [hunk 12]
```diff
@@ -151,4 +152,27 @@ public void testFailureClosesAllSubContexts() throws Exception {
...
+
+        AbstractExecutionSubContextTest.TestingExecutionSubContext ctx1 = new AbstractExecutionSubContextTest.TestingExecutionSubContext(1);
+        builder.addSubContext(ctx1);
+        AbstractExecutionSubContextTest.TestingExecutionSubContext ctx2 = new AbstractExecutionSubContextTest.TestingExecutionSubContext(2);
+        builder.addSubContext(ctx2);
+        JobExecutionContext jobExecutionContext = builder.build();
+
+        jobExecutionContext.start();
+        // fake execution time so we can sure the measurement is > 0
+        Thread.sleep(5L);
```
Comment (mxm) [inline sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java]: We could use a custom `TimeMeasurable` here which advances the time on start. Or maybe just decrease the sleep time. It's quite long.

File: sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    @Test
+    public void testExplainSelectWithoutJobExecutionContexts() {
+        execute("explain analyze select 1");
+        Map<String, Object> analysis = (Map<String, Object>) response.rows()[0][0];
+        Map<String, Object> executeAnalysis = (Map<String, Object>) analysis.get("Execute");
+        assertTrue(executeAnalysis.keySet().contains("Total"));
+        assertThat(executeAnalysis.keySet().size(), is(2)); // "Total" + local node id
```
Comment (mxm) [inline sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java]: Would be safer if we could really check the times here and use a `TimeMeasurable` which implements some fixed times.

File: core/src/main/java/io/crate/profile/ProfilingContext.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        return map.build();
+    }
+
+    public TimeMeasurable createAndStartMeasurable(String name) {
+        TimeMeasurable timer = createMeasurable(name);
+        timer.start();
+        return timer;
+    }
+
+    public void stopAndAddMeasurable(TimeMeasurable token) {
```
Comment (mxm) [inline core/src/main/java/io/crate/profile/ProfilingContext.java]: Still named token here.

File: core/src/main/java/io/crate/profile/ProfilingContext.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        token.stop();
+        if (enabled) {
+            map.put(token.name(), token.durationNanos() / 1_000_000L);
+        }
+    }
+
+    public TimeMeasurable createMeasurable(String name) {
+        return tokenProvider.apply(name);
```
Comment (mxm) [inline core/src/main/java/io/crate/profile/ProfilingContext.java]: Still named token here.

File: sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java (hunk 13 above)
Comment (chaudum) [reply sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java]: I'd rather not want to mock implementations in integration tests.

File: sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java (hunk 12 above)
Comment (chaudum) [reply sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java]: Hm, 5ms is not *that* long.
Being able to mock the TimeMeasurable that is used in the ProfilingContext in the JobExecutionContext would require quite some changes. Not sure if it would be worth the effort, tbh.

Comment (mxm) [reply sql/src/test/java/io/crate/execution/jobs/JobExecutionContextTest.java]: Sorry, for a second (pun intended) I forgot that Thread.sleep uses milliseconds.

File: sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java (hunk 13 above)
Comment (mxm) [reply sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java]: Could we check the node id here? We should have it available.

Comment (chaudum) [reply sql/src/test/java/io/crate/integrationtests/ExplainAnalyzeIntegrationTest.java]: sure

File: core/src/main/java/io/crate/profile/ProfilingContext.java (hunk 6 above)
Comment (andreidan) [reply core/src/main/java/io/crate/profile/ProfilingContext.java]: @mxm System.currentTimeMillis is not monotonic so it could lead to negative values. nanoTime _is_ monotonic on platforms where CLOCK_MONOTONIC is available

Comment (mxm) [reply core/src/main/java/io/crate/profile/ProfilingContext.java]: Thanks for the info @andreidan. Makes sense to use nanoTime then (also for precision reasons).
